### PR TITLE
Improve bounding box algorithms

### PR DESCRIPTION
**Pull request by permcody**
## Reason
Slow performance caused us to observe the behavior of what should have been a reasonably fast (O(n)) merge algorithm in the `PolycrystalUserObjectBase` class. Significant numbers of bounding boxes were being created for most features/grains.

## Design

The algorithm to build the bounding boxes was flawed, but did not produce an invalid result. The issue stemmed from merging sets by repeatedly merging overlapping bounding boxes. However, the algorithm didn't repeat this until full convergence due to an invalid assumption of the ordering of these bounding boxes. This resulted in far more bounding boxes than there should have been for most problems. 

A new algorithm now continues to merge boxes until no boxes are merged.

## Impact 

The final features will now have fewer bounding boxes (the minimum numbers) resulting in faster comparisons for merging, which will yield faster performance.

### CODE REVIEW

**Comment by permcody** [comment]
@dschwen - Ready for review: 

You might just look at the first commit by itself. The second is just upgrading all the for loops in this class to "index range". 

On Testing... I thought about this, but didn't come up with anything. We don't report the number of bounding boxes anywhere outside of these classes. I put in print statements and it's drastically down. Before this patch I ran into cases where some of the grains had hundreds of bounding boxes due to the existing faulty algorithm. With the new algorithm, I'm down to just single digit bounding boxes as one would expect. This should speed-up all of the intersection checks everywhere. I don't have anything quantitative. Perhaps in the future I could put in a few asserts if we get over the absolute maximum number of expected bounding boxes. This isn't trivial to figure out though. Usually it's related to the dimensionality and number of periodic boundary conditions, but if you support non-convex shapes, you can still end up with more bounding boxes than you might expect. The good news is that all the tests and examples appear to be running the same with these changes.

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C [hunk 1]
```diff
@@ -366,7 +366,7 @@ FeatureFloodCount::execute()
       if (elem->processor_id() == rank)
       {
         if (hasBoundary(boundary_id))
-          for (MooseIndex(_vars) var_num = 0; var_num < _vars.size(); ++var_num)
+          for (const auto var_num : make_range(_vars.size()))
```
Comment (dschwen) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C:369]:
```suggestion
          for (const auto var_num : index_range(_vars))
```

for future reference

**Review by dschwen — DISMISSED** [review]

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C (hunk 1 above)
Comment (permcody) [reply modules/phase_field/src/postprocessors/FeatureFloodCount.C:369]: A damn... I did this enough that it's probably worth fixing. I'll push those changes and also remove an unused function that I orphaned and then we can wrap this up. Thanks!

**Comment by permcody** [comment]
@dschwen - Once more with `index_range` where applicable, and a dead utility routine removed.

**Review by dschwen — APPROVED** [review]

**Pull request merged**